okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/DeprecationWarningInterceptor.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.okhttp;
+
+import com.palantir.logsafe.SafeArg;
+import java.io.IOException;
+import okhttp3.Interceptor;
+import okhttp3.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An interceptor that logs warnings when the response from a server contains the "deprecation" header. Logs include
+ * the content of the "server" header when it is provided, and always include the provided {@code serviceClassName} so
+ * that consumers applying this interceptor can locate the cause of the deprecation.
+ *
+ * <p>Note: endpoint information is not included because endpoint-level details are not available at this level.
+ */
+final class DeprecationWarningInterceptor implements Interceptor {
+    private static final Logger log = LoggerFactory.getLogger(DeprecationWarningInterceptor.class);
+    private final String serviceClassName;
+
+    private DeprecationWarningInterceptor(String serviceClassName) {
+        this.serviceClassName = serviceClassName;
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response = chain.proceed(chain.request());
+
+        if (response.isSuccessful()) {
+            String deprecationHeader = response.header("deprecation");
+            if (deprecationHeader != null) {
+                log.warn("Using a deprecated endpoint when connecting to service",
+                        SafeArg.of("serviceClass", serviceClassName),
+                        SafeArg.of("service", response.header("server", "no server header provided")));
```

Comment:
This warning will be logged on every request, thoughts on a rate limiter?
We should mark a metric each time.

Reply:
put in a once-per-minute rate limiter and added metrics